### PR TITLE
CORE-1276 memory APIs

### PR DESCRIPTION
**Pull request by mmogilnitsky**
total and free size.

### CODE REVIEW

File: modules/canvaskit/Makefile [hunk 1]
```diff
@@ -113,6 +112,32 @@ bluescape_debug:
 	cp ../../out/canvaskit_wasm_debug/canvaskit.js   ./npm_build-debug/bin
 	cp ../../out/canvaskit_wasm_debug/canvaskit.wasm ./npm_build-debug/bin
 
+bluescape_info:
+	# Does an incremental build where possible.
+	# Note that we shouldn't disable: no_fonts no_paragraph primitive_shaper
+	# rm -rf ../../out/canvaskit_wasm_info
```
Comment (ronpfeifle) [inline modules/canvaskit/Makefile]: hmmm.  did you intend to leave this comment line?

File: modules/canvaskit/compile.sh [hunk 2]
```diff
@@ -25,19 +25,34 @@ EMCC=`which emcc`
 EMCXX=`which em++`
 EMAR=`which emar`
 
+# Emscripten uses:
+# wasmMemory = new WebAssembly.Memory({ 'initial': ... , 'maximum': ...})
+# Unfortunately this buffer as well as its parameters are not directly
+# accessible, hence providing some extra API.
+INITIAL_MEMORY_MB=3
+MAXIMUM_MEMORY_MB=192
+MEMEORY_FLAGS="\"-DSK_INITIAL_MEMORY_MB=$INITIAL_MEMORY_MB\", \"-DSK_MAXIMUM_MEMORY_MB=$MAXIMUM_MEMORY_MB\", "
```
Comment (ronpfeifle) [inline modules/canvaskit/compile.sh]: typo 'MEMEORY'... too many E's

File: modules/canvaskit/compile.sh [hunk 3]
```diff
@@ -25,19 +25,34 @@ EMCC=`which emcc`
...
+
 RELEASE_CONF="-Oz --closure 1 -DSK_RELEASE --pre-js $BASE_DIR/release.js"
-EXTRA_CFLAGS="\"-DSK_RELEASE\","
+EXTRA_CFLAGS="\"-DSK_RELEASE\", $MEMEORY_FLAGS"
```
Comment (ronpfeifle) [inline modules/canvaskit/compile.sh]: typo here...

File: modules/canvaskit/compile.sh [hunk 4]
```diff
@@ -25,19 +25,34 @@ EMCC=`which emcc`
...
 IS_OFFICIAL_BUILD="true"
 
 IS_DEBUG="false"
 if [[ $@ == *debug* ]]; then
   echo "Building a Debug build"
-  EXTRA_CFLAGS="\"-DSK_DEBUG\","
+  EXTRA_CFLAGS="\"-DSK_DEBUG\", $MEMEORY_FLAGS"
```
Comment (ronpfeifle) [inline modules/canvaskit/compile.sh]: typo here

File: modules/canvaskit/compile.sh [hunk 5]
```diff
@@ -25,19 +25,34 @@ EMCC=`which emcc`
...
   RELEASE_CONF="-O0 --js-opts 0 -s DEMANGLE_SUPPORT=1 -s ASSERTIONS=1 -s GL_ASSERTIONS=1 -g3 \
                 --source-map-base /node_modules/canvaskit/bin/ -DSK_DEBUG --pre-js $BASE_DIR/debug.js"
   BUILD_DIR=${BUILD_DIR:="out/canvaskit_wasm_debug"}
   IS_DEBUG="true"
   IS_OFFICIAL_BUILD="false"
+elif [[ $@ == *info* ]]; then
+  echo "Building an Info build"
+  EXTRA_CFLAGS="\"-DSK_RELEASE\", $MEMEORY_FLAGS"
```
Comment (ronpfeifle) [inline modules/canvaskit/compile.sh]: typo here

File: modules/canvaskit/compile.sh [hunk 6]
```diff
@@ -25,19 +25,34 @@ EMCC=`which emcc`
...
+  RELEASE_CONF="-O0 --js-opts 0 -g3  \
+                --source-map-base /node_modules/canvaskit/bin/ -DSK_RELEASE --pre-js $BASE_DIR/debug.js"
```
Comment (ronpfeifle) [inline modules/canvaskit/compile.sh:53]: SK_RELEASE is right for the info build?  Just checking (I don't know myself)

File: include/private/SkMalloc.h [hunk 7]
```diff
@@ -140,4 +140,12 @@ static inline int sk_careful_memcmp(const void* a, const void* b, size_t len) {
...
 }
 
+SK_API extern unsigned int sk_get_initial_memory();
+
+SK_API extern unsigned int sk_get_maximum_memory();
+
+SK_API extern unsigned int sk_get_total_memory();
+
+SK_API extern unsigned int sk_get_free_memory();
+
```
Comment (ronpfeifle) [inline include/private/SkMalloc.h]: they all need a tiny comment to say what units are being used for memory sizes (ie, that it's in bytes, according to the .cpp file below)

Comment (leoyanggit) [inline include/private/SkMalloc.h]: Can you please move this declaration somewhere else, or potentially a new file? SkMalloc.h is a private header and this is no SK_API for the functions inside it. But the functions you add is SK_API.

File: src/ports/SkMemory_info.cpp [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+ * is allocated on WebAssembly.Memory(...). Yet this number is relative to
+ * the currently available getTotalMemory() heap. Thus:
+ *  currentHeapAllocated = sk_get_total_memory() - sk_get_free_memory()
+ *  maximalFreeMemory = sk_get_maximum_memory() - (sk_get_total_memory() - sk_get_free_memory())
+ */
+unsigned int sk_get_free_memory() {
+    s_mallinfo info = mallinfo();
+    unsigned int totalMemory = sk_get_total_memory();
+    unsigned int dynamicTop = (unsigned int)sbrk(0);
+    return totalMemory - dynamicTop + info.fordblks;
```
Comment (ronpfeifle) [inline src/ports/SkMemory_info.cpp:65]: okay, this is obviously the whole point of the PR.

I think there's a "units problem" here.  `totalMemory` and `info.fordblks` are counts of memory.  But
`dynamicTop = sbrk(0)` returns a memory address -- I don't understand how that fits with the counts.

**Review by ronpfeifle — COMMENTED** [review]
I'm confused about the most important 2 lines in the PR.

File: include/private/SkMalloc.h (hunk 7 above)
Comment (mmogilnitsky) [reply include/private/SkMalloc.h]: Don't be confused @leoyanggit by the word private. :) It doesn't seem to be a big of a deal in Skia ...
- There are SK_APIs in this file.
- private includes are called in canvas_bindings.cpp

Also the operations are very tightly related to various mallocs. So this is a very right place to declare them ....  in general. Thus we might be required to add another encapsulation just to make word private hidden., which I think is a bad idea.

So I think to leave it the way it is. At least for now. :)

File: modules/canvaskit/compile.sh (hunk 2 above)
Comment (mmogilnitsky) [reply modules/canvaskit/compile.sh]: :)

File: src/ports/SkMemory_info.cpp (hunk 8 above)
Comment (ronpfeifle) [reply src/ports/SkMemory_info.cpp:65]: I understand that this is in the WASM environment, so there could be some "monkey business" wrt to addresses.  But I'm still unclear about the implications of that.

Comment (mmogilnitsky) [reply src/ports/SkMemory_info.cpp:65]: https://github.com/emscripten-core/emscripten/blob/9bb322f8a7ee89d6ac67e828b9c7a7022ddf8de2/tests/mallinfo.cpp

File: modules/canvaskit/compile.sh (hunk 6 above)
Comment (mmogilnitsky) [reply modules/canvaskit/compile.sh:53]: Can not touch this.
The flag is all over the place and in other libraries. Notice that it is also used in DEBUG case above.

File: include/private/SkMalloc.h (hunk 7 above)
Comment (mmogilnitsky) [reply include/private/SkMalloc.h]: initial_memory and maximum_memory are EMscripten's notation the rest is simply derived.

File: src/ports/SkMemory_info.cpp (hunk 8 above)
Comment (mmogilnitsky) [reply src/ports/SkMemory_info.cpp:65]: From above: 
 * Emscripten uses:
 * wasmMemory = new WebAssembly.Memory({ 'initial': ... , 'maximum': ...})
 * Unfortunately this buffer as well as its parameters are not directly
 * accessible, hence providing some extra API.

@ronpfeifle it follows that we have all our HEAP8 as a buffer with adresses of:
0 ... HEAP8.length
Hence to get the current chunk one needs to:
totalMemory - dynamicTop

Comment (ronpfeifle) [reply src/ports/SkMemory_info.cpp:65]: Ah -- okay, that's just the way WASM works, I guess.  Thanks for the explanation! 😄

**Review by ronpfeifle — APPROVED** [review]
well.... seems like it should be okay.

File: include/private/SkMalloc.h (hunk 7 above)
Comment (mmogilnitsky) [reply include/private/SkMalloc.h]: Ohh sorry merged it @leoyanggit 
Let me know if you would like it other way around and I will revert.

**Pull request merged**